removed circular dependency

JanHuss opened the pull request:
Issue: Circular dependency was preventing the application to run

Solve: Removed one of the `Include` from the header

JanHuss committed:
removed circular dependency

JanHuss committed:
made a few changes, mouse input, game reset, paddle in boundries

JanHuss committed:
added some colour to the bricks

JanHuss committed:
added level and game loop

JanHuss committed:
tried to get the bricks to change colour when colliding with ball.

JanHuss commented:
Added features: 
- Game Loop for game over and Level complete (would preferred to setup state machine)
- Changed colours on bricks (was trying to give bricks different strengths)
- Mouse input (couldn't win without it, also added boundaries to the paddle based on the size of the window

david-uad commented on Breakout/GameManager.cpp:
```diff
@@ -77,9 +112,28 @@ void GameManager::update(float dt)
+	if (localMousePosition.x + _paddle->getWidth()/2 > _window->getSize().x) localMousePosition.x = _window->getSize().x - 0.5f - _paddle->getWidth()/2;
+
+    // move paddle with mouse
+    _paddle->setPosition(static_cast<float>(localMousePosition.x)); 
```
The only issue here is that this method of mouse input does not account for the predefined speed of the paddle defined by its class and the pickups.

david-uad commented on Breakout/GameManager.cpp:
```diff
@@ -110,6 +165,56 @@ void GameManager::levelComplete()
+void GameManager::resetGame(std::string message, int levelIncrease)
+{
+	// game over
+    if (_masterText.getString() == gameOverText)
```
There is some reuse of code in each condition here. This could be either abstracted into two different functions for each game state and one to reset the common member states, or refactored so that the conditional statement only contains the difference in code, with the rest executed in the function outwith the conditional, regardless of the game state.

JanHuss replied:
Totally agree with that. It's been on my mind so thanks for confirming

david-uad reviewed (commented):
The code is well readable, with verbatim naming for new members, and clear purposes for each new line. It is maintainable with no hardcoded values. It performs well and makes good use of resources. The coding standards and naming conventions have been adhered to well. The code is clear, consise and well commented.

The resetGame function could be refactored as the message parameter is unused, and a large portion of the code in the contained conditional is the same.

JanHuss committed:
added the engine to the sfml project

- playing the game tune within game manager class
- written down process of where the game needs to include the sound triggers: `ball` and `gameManager` classes

JanHuss committed:
removed audio assets

contained test assets

JanHuss committed:
added paddle ball sound

yaaaaaaaaaaaa

JanHuss committed:
added further audio assets to the game

- assets are playable and audible in various classes.
- easy to implement into SFML

JanHuss committed:
updated leaf class

JanHuss committed:
small changes

JanHuss committed:
did some debugging for the audio to work

audio engine works no issues. but the breakout game itself needs a lot of work. might try and refactor some of the game manager code over the weekend

JanHuss committed:
added 2 further audio assets to game

- made a better game over sound
- created a level completed sound

JanHuss committed:
implemented the level complete and fixed a bug

- fixed having the game still be playable when game over or level completed. fix: set the pause value to a constant in the constant class

JanHuss committed:
made small changes for the game to have sounds play better

JanHuss committed:
updated audio engine files

JanHuss committed:
cleaned up project

updated the filters in game and am currently adding a statemachine within the game manager class

JanHuss committed:
added features

- cleaned up game manager
- state machine is in place within game manager. now I can find things easier.
- set the paddle to increase gradually rather than instantly. little issue is that the paddle's centre position is not adjusting itself. will need to calculate it to recentre itself properly.
- ball now has sounds for the boundries. was attempting to use the same asset of the paddle sound but apply pitch features instead. this didn't work as expected and the sound stays in the same pitch. it was working within the audio engine. so would be good to see what is so different there.
would like to get the paddle calculations sorted next so that this topic is cleared and can move onto next power up.

JanHuss committed:
added further sounds and updated the engine

JanHuss committed:
tried to fix eventmanager

JanHuss committed:
added pitch and panning to powerups and boundries

- added pitch and panning to powerups and boundries
- also amended a bit of the mouse input to create the boundries but its still actiing odd on initialisation
- fixed paddle expand and shrinking

JanHuss committed:
amended input controls

- using keyboard over mouse for input. is a bit more fun imo. also, using mouse was causing a clash with the keyboard controls.
- moved keyboard controls to paddle class as that is where all player input happens

JanHuss committed:
started working on brick manager and brick lives

- is rendering the colours correctly but needs some work when ball collides with brick to set life

JanHuss committed:
fixed bug

- brickmanager uses a counter to iterate through all bricks and that counter wasn't being reset so i was getting an out of bounds error
- also started looking into when the ball is reset in game manager and what I could potentially do to create a little pre timer before the game starts

JanHuss committed:
added two sprites to follow the main ball when the fast ball is active

JanHuss committed:
added a ball manager

- there was no way around it and a ball manager needed to be implemented
- the reason: the ball needs a manager now that multiball and double effects have to be implemented.
- there might be more effects later that would benefit from a ball management system

JanHuss committed:
moved ball parameter from power up manager

- ball has been removed from game manager and placed within ball manager.
- the game can run without issues, as long as the paddle does not interact specifically with a ball powerup.
- this is because the ball has still got an expected nullptr for moving around the code.
- the idea is to apply the effects via ballmanager instead of the ball itself. however, the ball class contains a lot of parameters needed within the ball class or so I believe.
- I might be able to iterate throught the vector of balls itself within the ballmanager class and apply the powerups within the ball class as well. however, I would like ball to only contain its movement functionality and any parameter changes it receives from the ball manager. some food for thought.

JanHuss committed:
fixed nullptr powerup bug

- directed all powerups to look at balls vector in ball manager instead of ball class
- removed all ball pointers within powerups as no longer needed

JanHuss committed:
tidied up code after amending ball powerup nullptr bug

JanHuss committed:
added extra ball class called powerup double ball.

- was going to have two separate powerups with similar capabilities but thought of a good way to make it more interesting. if the player manages to get more than two balls then the player will stay immune when loosing a ball.

JanHuss committed:
random ball restart

added a little random range to the ball when restarting but I think I will have to amend this a little
